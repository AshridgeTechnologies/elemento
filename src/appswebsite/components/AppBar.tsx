import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}