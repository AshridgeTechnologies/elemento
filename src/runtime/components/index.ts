export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
