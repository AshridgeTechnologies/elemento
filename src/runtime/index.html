<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>