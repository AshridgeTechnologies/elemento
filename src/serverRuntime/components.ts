export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts