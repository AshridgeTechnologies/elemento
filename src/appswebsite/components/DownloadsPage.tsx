import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
