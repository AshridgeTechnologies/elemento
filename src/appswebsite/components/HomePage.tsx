import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import HomeIcon from '@mui/icons-material/Home'
import LightbulbIcon from '@mui/icons-material/LightbulbTwoTone'
import CreateIcon from '@mui/icons-material/CreateTwoTone'
import AodIcon from '@mui/icons-material/AodTwoTone'
import ArrowCircleRightTwoToneIcon from '@mui/icons-material/ArrowCircleRightTwoTone';

const primary = theme.palette.primary.main
const secondary = theme.palette.secondary.main

function Logo() {
    return <Box color={secondary} border='8px solid' borderRadius={4} width={108} height={104} paddingTop='10px' paddingX='3px'>
        <Stack>
            <Typography fontSize='100px' fontFamily='"Helvetica"' align='center' lineHeight={0.8}>El</Typography>
            <Typography fontSize='16px' lineHeight={1}>42</Typography>
        </Stack>
    </Box>
}

function SectionLink(props: {title: string, href: string, icon: any, children: any}) {
    const Icon = props.icon
    return <Box border='4px solid' borderColor={primary} borderRadius={4} width='300px' paddingY='10px' paddingX='16px'>
        <Link href={props.href} underline='none'>
            <Stack spacing={2}>
                <Box textAlign='center'><Icon sx={{ fontSize: 180 }} /></Box>
                <Link component='div' underline='hover' href={props.href} variant='h5' textAlign='center'>{props.title}</Link>
                <Typography fontSize='20px' color='#555'>{props.children}</Typography>
                <Box textAlign='center'><ArrowCircleRightTwoToneIcon sx={{ fontSize: 32, color: secondary }} /></Box>
            </Stack>
        </Link>
    </Box>
}

export default function HomePage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Home'/>
            </Box>
            <Box flex='1' minHeight={0} mt={2}>
                <Container maxWidth="lg">
                    <Stack spacing={2}>
                        <Stack direction='row'>
                            <Logo/>
                            <Stack ml={4}>
                                <Typography fontWeight='bold' variant='h3'  color={secondary} align='center'>Elemento</Typography>
                                <Typography variant='h4' color={secondary} align='center' mt={2}>
                                    Everything you need to make and use your own software
                                </Typography>
                            </Stack>
                        </Stack>
                        <Stack direction='row' justifyContent='space-between' alignItems='stretch' marginTop='30px'>
                            <SectionLink title='Learn' href='/learn' icon={LightbulbIcon}>
                                A great range of resources to teach you how to get started and create your first app with Elemento.
                            </SectionLink>
                            <SectionLink title='Create' href='/studio' icon={CreateIcon}>
                                Put your app together and preview it live in Elemento Studio
                            </SectionLink>
                            <SectionLink title='Use' href='/run' icon={AodIcon}>
                                Run your app - or go back to the apps you have already made.
                            </SectionLink>
                        </Stack>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
