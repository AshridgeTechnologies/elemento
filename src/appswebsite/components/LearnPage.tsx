import React from 'react'
import Toolbar from '@mui/material/Toolbar'
import {Box, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import {TutorialRunner} from '../../tutorials/TutorialRunner'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Learn' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <TutorialRunner/>
            </Box>
        </Box>
    </ThemeProvider>
}
