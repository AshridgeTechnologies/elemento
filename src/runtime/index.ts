import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
