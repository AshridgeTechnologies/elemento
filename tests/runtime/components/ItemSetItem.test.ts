import {test} from "vitest"
/**
 * @vitest-environment jsdom
 */
import React, {createElement} from 'react'
import {ItemSetItem, TextElement} from '../../../src/runtime/components'
import {inDndContext, snapshot, valueObj} from '../../testutil/testHelpers'

test('ItemSetItem produces output containing ReactElement children', () => {
    const childEl = createElement(TextElement, {path: 'item1', content: 'where are you?'})
    const styles = {color: 'red', borderWidth: 20, backgroundColor: valueObj('blue')}
    const itemElement: React.ReactElement = createElement(ItemSetItem, {
        path: 'page1.para1', item: 'red', index: 3, itemId: '3', onClick: () => console.log('Click'), styles, children: childEl})
    const elementInDndContext = inDndContext(itemElement)
    snapshot(elementInDndContext)()
})

test('ItemSetItem produces output containing ReactElement children if draggable', () => {
    const childEl = createElement(TextElement, {path: 'item1', content: 'where are you?'})
    const styles = {color: 'red', borderWidth: 20, backgroundColor: valueObj('blue')}
    const itemElement: React.ReactElement = createElement(ItemSetItem, {
        path: 'page1.para1', item: 'red', index: 3, itemId: '3', onClick: () => console.log('Click'), canDragItem: true, styles, children: childEl})
    const elementInDndContext = inDndContext(itemElement)
    snapshot(elementInDndContext)()
})

test('ItemSetItem produces output containing ReactElement children is draggable is a false value object', () => {
    const childEl = createElement(TextElement, {path: 'item1', content: 'where are you?'})
    const styles = {color: 'red', borderWidth: 20, backgroundColor: valueObj('blue')}
    const itemElement: React.ReactElement = createElement(ItemSetItem, {
        path: 'page1.para1', item: 'red', index: 3, itemId: '3', onClick: () => console.log('Click'), canDragItem: valueObj(false), styles, children: childEl})
    const elementInDndContext = inDndContext(itemElement)
    snapshot(elementInDndContext)()
})
