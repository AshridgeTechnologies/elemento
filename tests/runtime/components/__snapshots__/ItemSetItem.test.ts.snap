// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSetItem produces output containing ReactElement children 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-0"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1frel66"
    id="page1.para1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      id="item1"
      style={{}}
    >
      where are you?
    </div>
  </div>,
  <div
    id="DndDescribedBy-0"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`ItemSetItem produces output containing ReactElement children if draggable 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-1"
    aria-disabled={false}
    aria-roledescription="draggable"
    className="MuiBox-root css-qls7rp"
    id="page1.para1"
    onClick={[Function]}
    onPointerDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      id="item1"
      style={{}}
    >
      where are you?
    </div>
  </div>,
  <div
    id="DndDescribedBy-1"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-1"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`ItemSetItem produces output containing ReactElement children is draggable is a false value object 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-2"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-qls7rp"
    id="page1.para1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      id="item1"
      style={{}}
    >
      where are you?
    </div>
  </div>,
  <div
    id="DndDescribedBy-2"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-2"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;
